File: grind75/week1/042-Insert-Interval-Medium/042-Insert-Interval-Medium.py
# Question link: https://leetcode.com/problems/insert-interval/description/

# In place solution is not required for this problem, but I'll still include it as a possible solution


 
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:

        # edge case
        if not intervals:
            return [newInterval]
        
        start = 0
        end = 1
        res = []

        for i in range(len(intervals)):
            # new interval is to the right of this current interval
            if intervals[i][end] < newInterval[start]:
                res.append(intervals[i])

            # new interval is to the left of this current interval
            elif intervals[i][start] > newInterval[end]:
                res.append(newInterval)
                return res + intervals[i:]

            # there's an overlap
            else:
                newInterval[start] = min(intervals[i][start], newInterval[start])
                newInterval[end] = max(intervals[i][end], newInterval[end])

        # accounts for edges like this:
        # example 1: 
        # .---.
        #   .--new--.
        # example2:
        # .---.
        #       .---new---.
        res.append(newInterval)

        return res

                

# In place solution

class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        start = i = 0
        end = 1
        while i < len(intervals):
            if newInterval[start] <= intervals[i][end]:
                if newInterval[end] < intervals[i][start]:
                    break
                newInterval[start] = min(newInterval[start], intervals[i][start])
                newInterval[end] = max(newInterval[1], intervals[i][end])
                intervals.pop(i)
            else:
                i += 1
        intervals.insert(i, newInterval)
        return intervals

        


        

File: grind75/week1/042-Insert-Interval-Medium/042-Insert-Interval-Medium.cpp
// Problem link: https://leetcode.com/problems/insert-interval/

 class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        // Edge case: if intervals are empty
        if (intervals.empty()) {
            return { newInterval };
        }

        const int start = 0;
        const int end = 1;
        vector<vector<int>> res;

        for (int i = 0; i < intervals.size(); i++) {
            // New interval is to the right of the current interval
            if (intervals[i][end] < newInterval[start]) {
                res.push_back(intervals[i]);
            } 
            // New interval is to the left of the current interval
            else if (intervals[i][start] > newInterval[end]) {
                res.push_back(newInterval);
                res.insert(res.end(), intervals.begin() + i, intervals.end());
                return res;
            } 
            // There's an overlap
            else {
                newInterval[start] = min(intervals[i][start], newInterval[start]);
                newInterval[end] = max(intervals[i][end], newInterval[end]);
            }
        }

        // Accounts for edge cases
        res.push_back(newInterval);

        return res;
    }
};
